### PR TITLE
[Torch-Dialect] add canonicalizer for prim::ListConstruct op

### PR DESCRIPTION
**Pull request by Vremold**

### CODE REVIEW

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 1]
```diff
@@ -2085,6 +2085,36 @@ void PrimTupleUnpackOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
+void PrimListConstructOp::getCanonicalizationPatterns(
+    RewritePatternSet &patterns, MLIRContext *context) {
+  patterns.add(+[](PrimListConstructOp op, PatternRewriter &rewriter) {
+    if (isListPotentiallyMutated(op.getResult())) 
+      return failure();
+    SmallVector<Value> elements = llvm::to_vector<4>(op.getElements());
+    if (elements.size() == 0)
+      return failure();
+
+    auto listUnpackOp = elements[0].getDefiningOp<PrimListUnpackOp>();
```
Comment (mgehre-amd) [inline lib/Dialect/Torch/IR/TorchOps.cpp:2101]: Why do we only need to check the first element?

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 2]
```diff
@@ -2085,6 +2085,36 @@ void PrimTupleUnpackOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
...
+    if (!listUnpackOp)
+      return failure();
+    SmallVector<Value, 4> listUnpackResults;
+    for (auto v : listUnpackOp.getResults()) {
+      listUnpackResults.emplace_back(v);
```
Comment (mgehre-amd) [inline lib/Dialect/Torch/IR/TorchOps.cpp:2106]: Instead of adding those into `listUnpackResults`, can we directly compare against `elements`?

Comment (Vremold) [reply lib/Dialect/Torch/IR/TorchOps.cpp]: Good idea! The new patch fixes this.

File: test/Dialect/Torch/canonicalize.mlir [hunk 3]
```diff
@@ -965,6 +965,15 @@ func.func @torch.prim.If$fold_same_result$subset_of_results(%arg0: !torch.bool,
...
 }
 
+// CHECK-LABEL: func.func @prim.ListConstruct$fold_list(
+// CHECK-SAME:      %[[ARG0:.*]]: !torch.list<tensor>) -> !torch.list<tensor> {
+// CHECK:         return %[[ARG0]] : !torch.list<tensor>
+func.func @prim.ListConstruct$fold_list(%arg0: !torch.list<tensor>) -> !torch.list<tensor> {
+  %0:2 = torch.prim.ListUnpack %arg0 : !torch.list<tensor> -> !torch.tensor, !torch.tensor
+  %1 = torch.prim.ListConstruct %0#0, %0#1 : (!torch.tensor, !torch.tensor) -> !torch.list<tensor>
+  return %1 : !torch.list<tensor>
+}
```
Comment (qingyunqu) [inline test/Dialect/Torch/canonicalize.mlir:975]: I think it's necessary to check that `%arg0`'s length is same as `NumResults` of `prim.Listunpack`.

Comment (Vremold) [reply test/Dialect/Torch/canonicalize.mlir:975]: We only replace the later ListConstruct op with the former ListUnpack's operand. ListUnpack op will not be eliminated.

**Comment by Vremold** [comment]
Could you review the PR for me? @ramiro050

**Review by mgehre-amd — APPROVED** [review]
LGTM

**Pull request merged**